File: sorting/rust/bubble_sort.rs
fn bubble_sort(arr: &mut [i32]) {
    let n: usize = arr.len();

    for i in (0..n).rev() {
        let mut swapped: bool = false;

        for j in 0..i {
            if arr[j] > arr[j + 1] {
                arr.swap(j, j + 1);
                swapped = true;
            }
        }

        if !swapped {
            break;
        }
    }
}

fn main() {
    let mut arr: [i32; 20] = [38, 27, 43, 3, 9, 82, 10, 1, 75, 56, 14, 62, 48, 91, 23, 7, 55, 30, 18, 99];

    bubble_sort(&mut arr);
    println!("{:?}", arr);
}

File: algorithms/sorting/bubble_sort.py
def bubble_sort(arr: list[int]) -> list[int]:
    arr: list[int] = arr.copy()
    n: int = len(arr)

    for i in reversed(range(0, n)):
        swapped = False

        for j in range(0, i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True

        if not swapped:
            break

    return arr


if __name__ == "__main__":
    arr: list[int] = [
        38,
        27,
        43,
        3,
        9,
        82,
        10,
        1,
        75,
        56,
        14,
        62,
        48,
        91,
        23,
        7,
        55,
        30,
        18,
        99,
    ]

    print(bubble_sort(arr))
